### DISCUSSION TITLE
Allow xcom_pull to return multiple results if task_ids is None

### DISCUSSION THREAD

**Discussion by paavanb** [apache/airflow#23989] [Ideas]
**Description**

`TaskInstance.xcom_pull` only returns multiple xcom values if an array of task ids are passed in. This seems to make it impossible to get all XComs for a given key without also knowing exactly which `task_ids` set them, even though the documentation suggests that task ids are not necessary and that supplying `None` "removes the filter" ([comment here](https://github.com/apache/airflow/blob/1.10.12/airflow/models/taskinstance.py#L1540)). Is this intentional behavior?

Fortunately, I can work around this by writing my own function to call `XCom.get_many` directly, but it's not ideal.

**Use case / motivation**

Suppose there is a DAG in which tasks can declare artifacts that they output; these are all pushed as XComs with the same key (e.g., `"output-artifact"`). Future tasks attempting to use these artifacts can check for DAG consistency by pulling all XComs with that key and ensuring that a desired artifact was output by some previous task in the current DAG. But since these tasks are reusable/defined elsewhere, they don't know the `task_ids` of all preceding tasks. If `task_ids` is set to `None`, it seems that `xcom_pull` should return all XComs that match the parameters provided, but that's not true today.

Appreciate any feedback!

**Related Issues**

https://github.com/apache/airflow/pull/6461
https://issues.apache.org/jira/browse/AIRFLOW-5804

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by uranusjr** [comment]
This makes sense, but I don’t think this is viable without breaking _a lot_ of existing code? It’s difficult to tell how many existing code relies on this implicit behaviour.

**Comment by potiuk** [comment]
it also described below in main where task mapping is used (thouhg `task_id` misses the plural form). https://github.com/apache/airflow/blob/841ed271017ff35a3124f1d1a53a5c74730fed60/airflow/models/taskinstance.py#L2357 

I'd say much better solution will be to use a special object that you can pass to xcom_pull to get what you want more explicitly (and this is a chance to improve docs and remove ambiguities):

```
task_ids=TaskInstance.ALL_TASKS
```

Would you like to add a PR for that @paavanb  (unless @uranusjr  objects)
